### PR TITLE
Basic implementation for Volpiano

### PR DESCRIPTION
**Pull request by lpugin**
* Input option has to be `--input-from volpiano` (not auto detect)
* Input file is a single line text file
* Output in a single measure with stemless quarter notes
* Liquescents are encoded as `note@grace="unknown"
* Accidentals are encoded as `note@accid`
* Neume spacing / grouping is ignored
* System and page indications and bar lines are ignored
* Clef is always `G-2`

Example:
```
1---dc--d---feD---dcd---fef--g--hijhgf--gf---3---fIed
```

image

<details>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0">
   <meiHead>
      <fileDesc>
         <titleStmt>
            <title />
         </titleStmt>
         <pubStmt>
            <date isodate="2024-07-22-11:55:37" />
         </pubStmt>
      </fileDesc>
      <encodingDesc>
         <appInfo>
            <application xml:id="verovio" version="4.3.0-dev-0efb136">
               <name>Verovio (4.3.0-dev-0efb136)</name>
            </application>
         </appInfo>
         <projectDesc>
            <p>MEI encoded with Verovio</p>
         </projectDesc>
      </encodingDesc>
   </meiHead>
   <music>
      <body>
         <mdiv xml:id="m60fw6p">
            <score xml:id="s2gvssn">
               <scoreDef xml:id="s8it5ki">
                  <staffGrp xml:id="skqhuru">
                     <staffDef xml:id="sa4ycu2" n="1" lines="5" clef.shape="G" clef.line="2" />
                  </staffGrp>
               </scoreDef>
               <section xml:id="s1ikraa1">
                  <measure xml:id="m12bon0y">
                     <staff xml:id="soztiyy" n="1">
                        <layer xml:id="l1vmh78" n="1">
                           <note xml:id="n1hpak6b" dur="4" oct="4" pname="d" stem.len="0.000000" />
                           <note xml:id="nfbm4pd" dur="4" oct="4" pname="c" stem.len="0.000000" />
                           <note xml:id="nq4to4g" dur="4" oct="4" pname="d" stem.len="0.000000" />
                           <note xml:id="n1m38fjf" dur="4" oct="4" pname="f" stem.len="0.000000" />
                           <note xml:id="n7imrvh" dur="4" oct="4" pname="e" stem.len="0.000000" />
                           <note xml:id="n1wzfc2a" dur="4" oct="4" pname="d" grace="unknown" stem.len="0.000000" />
                           <note xml:id="n7hlbum" dur="4" oct="4" pname="d" stem.len="0.000000" />
                           <note xml:id="n7qxvog" dur="4" oct="4" pname="c" stem.len="0.000000" />
                           <note xml:id="ntn6vde" dur="4" oct="4" pname="d" stem.len="0.000000" />
                           <note xml:id="n1v6uxzv" dur="4" oct="4" pname="f" stem.len="0.000000" />
                           <note xml:id="n11mxwgw" dur="4" oct="4" pname="e" stem.len="0.000000" />
                           <note xml:id="n11tzl54" dur="4" oct="4" pname="f" stem.len="0.000000" />
                           <note xml:id="nieumiu" dur="4" oct="4" pname="g" stem.len="0.000000" />
                           <note xml:id="n12r50jb" dur="4" oct="4" pname="a" stem.len="0.000000" />
                           <note xml:id="n61g3hs" dur="4" oct="4" pname="b" stem.len="0.000000" accid="f" />
                           <note xml:id="na7d6ja" dur="4" oct="4" pname="a" stem.len="0.000000" />
                           <note xml:id="njqz7dn" dur="4" oct="4" pname="g" stem.len="0.000000" />
                           <note xml:id="n7tpuzu" dur="4" oct="4" pname="f" stem.len="0.000000" />
                           <note xml:id="n5aeiwu" dur="4" oct="4" pname="g" stem.len="0.000000" />
                           <note xml:id="no9t1l9" dur="4" oct="4" pname="f" stem.len="0.000000" />
                           <note xml:id="n6twial" dur="4" oct="4" pname="f" stem.len="0.000000" />
                           <note xml:id="ntrlb2u" dur="4" oct="4" pname="e" stem.len="0.000000" accid="n" />
                           <note xml:id="n1suw446" dur="4" oct="4" pname="d" stem.len="0.000000" />
                        </layer>
                     </staff>
                  </measure>
               </section>
            </score>
         </mdiv>
      </body>
   </music>
</mei>
```

</details>

### CODE REVIEW

**Comment by ahankinson** [comment]
Perhaps remove `measure` so that it's `section` / `staff` / `layer`?

**Comment by lpugin** [comment]
image

(That removes the barline at the end)

<details>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0">
   <meiHead>
      <fileDesc>
         <titleStmt>
            <title />
         </titleStmt>
         <pubStmt>
            <date isodate="2024-07-22-12:35:54" />
         </pubStmt>
      </fileDesc>
      <encodingDesc>
         <appInfo>
            <application xml:id="verovio" version="4.3.0-dev-0efb136-dirty">
               <name>Verovio (4.3.0-dev-0efb136-dirty)</name>
            </application>
         </appInfo>
         <projectDesc>
            <p>MEI encoded with Verovio</p>
         </projectDesc>
      </encodingDesc>
   </meiHead>
   <music>
      <body>
         <mdiv xml:id="m1ov765s">
            <score xml:id="s1toqaq4">
               <scoreDef xml:id="stts8zs">
                  <staffGrp xml:id="snu0cj7">
                     <staffDef xml:id="s1da41yo" n="1" lines="5" clef.shape="G" clef.line="2" />
                  </staffGrp>
               </scoreDef>
               <section xml:id="s1mkitan">
                  <staff xml:id="s1et63yn" n="1">
                     <layer xml:id="lynbbid" n="1">
                        <note xml:id="nbjll2u" dur="4" oct="4" pname="d" stem.len="0.000000" />
                        <note xml:id="n119ylsn" dur="4" oct="4" pname="c" stem.len="0.000000" />
                        <note xml:id="n5k9izr" dur="4" oct="4" pname="d" stem.len="0.000000" />
                        <note xml:id="nfx30cn" dur="4" oct="4" pname="f" stem.len="0.000000" />
                        <note xml:id="nuzdb8" dur="4" oct="4" pname="e" stem.len="0.000000" />
                        <note xml:id="n7pw3c4" dur="4" oct="4" pname="d" grace="unknown" stem.len="0.000000" />
                        <note xml:id="nhonhjp" dur="4" oct="4" pname="d" stem.len="0.000000" />
                        <note xml:id="n1fcxry5" dur="4" oct="4" pname="c" stem.len="0.000000" />
                        <note xml:id="n112ll6k" dur="4" oct="4" pname="d" stem.len="0.000000" />
                        <note xml:id="n1cg5sxg" dur="4" oct="4" pname="f" stem.len="0.000000" />
                        <note xml:id="n2bvt4v" dur="4" oct="4" pname="e" stem.len="0.000000" />
                        <note xml:id="ncep8cm" dur="4" oct="4" pname="f" stem.len="0.000000" />
                        <note xml:id="n1yefaxf" dur="4" oct="4" pname="g" stem.len="0.000000" />
                        <note xml:id="n1qqlk4m" dur="4" oct="4" pname="a" stem.len="0.000000" />
                        <note xml:id="n3x677a" dur="4" oct="4" pname="b" stem.len="0.000000" accid="f" />
                        <note xml:id="n1deb1sj" dur="4" oct="4" pname="a" stem.len="0.000000" />
                        <note xml:id="n1vb48mw" dur="4" oct="4" pname="g" stem.len="0.000000" />
                        <note xml:id="n1aw9yg1" dur="4" oct="4" pname="f" stem.len="0.000000" />
                        <note xml:id="nze3o9y" dur="4" oct="4" pname="g" stem.len="0.000000" />
                        <note xml:id="n17evphu" dur="4" oct="4" pname="f" stem.len="0.000000" />
                        <note xml:id="n19d7ti5" dur="4" oct="4" pname="f" stem.len="0.000000" />
                        <note xml:id="nfz00pn" dur="4" oct="4" pname="e" stem.len="0.000000" accid="n" />
                        <note xml:id="n1fsbu7e" dur="4" oct="4" pname="d" stem.len="0.000000" />
                     </layer>
                  </staff>
               </section>
            </score>
         </mdiv>
      </body>
   </music>
</mei>
```
</details>

**Comment by ahankinson** [comment]
`stem.len="0.000000"` -- does this need so many significant digits? Would `stem.len="0"` be possible?

**Comment by ahankinson** [comment]
`grace="unknown"` -- Would `@fontsize="small"` be possible instead?

**Comment by craigsapp** [comment]
Yes,  numbers such as 0.00000 are probably created by using `to_string()`.   Using `stringstream` would avoid the trailing zeros for fractional numbers.  (Note: `value.str("")` can be used to clear the previous contents of the `stringstream` when using in a loop (there is also `stringstream::clear()` but this is for clearing error flags and not to clear the string buffer.

**Comment by ahankinson** [comment]
`@dur="4"`... Can this be removed? The notes have no fixed duration. If necessary, `@head.fill="solid"` can ensure they have the correct appearance.

**Comment by lpugin** [comment]
```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0">
   <meiHead>
      <fileDesc>
         <titleStmt>
            <title />
         </titleStmt>
         <pubStmt>
            <date isodate="2024-07-22-13:03:25" />
         </pubStmt>
      </fileDesc>
      <encodingDesc>
         <appInfo>
            <application xml:id="verovio" version="4.3.0-dev-8f2eb54-dirty">
               <name>Verovio (4.3.0-dev-8f2eb54-dirty)</name>
            </application>
         </appInfo>
         <projectDesc>
            <p>MEI encoded with Verovio</p>
         </projectDesc>
      </encodingDesc>
   </meiHead>
   <music>
      <body>
         <mdiv xml:id="m17tsv1h">
            <score xml:id="s2z46l2">
               <scoreDef xml:id="strkohc">
                  <staffGrp xml:id="s1xmlnzu">
                     <staffDef xml:id="s1rl39ji" n="1" lines="5" clef.shape="G" clef.line="2" />
                  </staffGrp>
               </scoreDef>
               <section xml:id="sndt6ny">
                  <staff xml:id="s1dpdqy5" n="1">
                     <layer xml:id="l132ldwq" n="1">
                        <note xml:id="nemwjjs" oct="4" pname="d" />
                        <note xml:id="nsqsefm" oct="4" pname="c" />
                        <note xml:id="n28h8g3" oct="4" pname="d" />
                        <note xml:id="n1a8jmdt" oct="4" pname="f" />
                        <note xml:id="nd546gf" oct="4" pname="e" />
                        <note xml:id="nyjnlka" oct="4" pname="d" cue="true" />
                        <note xml:id="n1si7q4s" oct="4" pname="d" />
                        <note xml:id="nhx8bg6" oct="4" pname="c" />
                        <note xml:id="nt8isna" oct="4" pname="d" />
                        <note xml:id="ns55ypj" oct="4" pname="f" />
                        <note xml:id="n3rd3uj" oct="4" pname="e" />
                        <note xml:id="n1xw729z" oct="4" pname="f" />
                        <note xml:id="nmfmm9" oct="4" pname="g" />
                        <note xml:id="n12u4ihq" oct="4" pname="a" />
                        <note xml:id="ngkpns1" oct="4" pname="b" accid="f" />
                        <note xml:id="nc916ce" oct="4" pname="a" />
                        <note xml:id="n17vuvxi" oct="4" pname="g" />
                        <note xml:id="n190crnw" oct="4" pname="f" />
                        <note xml:id="n1cyg6ay" oct="4" pname="g" />
                        <note xml:id="nqyibul" oct="4" pname="f" />
                        <note xml:id="n1uhx4no" oct="4" pname="f" />
                        <note xml:id="nloqva4" oct="4" pname="e" accid="n" />
                        <note xml:id="nq8qmv2" oct="4" pname="d" />
                     </layer>
                  </staff>
               </section>
            </score>
         </mdiv>
      </body>
   </music>
</mei>
```

File: src/iovolpiano.cpp [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    // add minimal scoreDef
+    StaffGrp *staffGrp = new StaffGrp();
+    StaffDef *staffDef = new StaffDef();
+    staffDef->SetN(1);
+    staffDef->SetLines(5);
+    // Arbitray clef
+    Clef *clef = new Clef();
+    clef->IsAttribute(true);
```
Comment (rettinghaus) [inline src/iovolpiano.cpp:121]: Even if we don't care for it for now, the clef _is_ encoded, so I'd prefer to have it as element:
```suggestion
    clef->IsAttribute(false);
```

**Review by rettinghaus — CHANGES_REQUESTED** [review]
Looks great. Two small things.

File: src/iovolpiano.cpp (hunk 1 above)
Comment (lpugin) [reply src/iovolpiano.cpp:121]: I think an attribute is fine for now.

File: src/iovolpiano.cpp [hunk 2]
```diff
@@ -103,10 +104,10 @@ bool VolpianoInput::Import(const std::string &volpiano)
             }
             layer->AddChild(note);
         }
-        else if (ch == 'y' || ch == 'i' || ch == 'z') {
+        else if (ch == 'i' || ch == 'w' || ch == 'x' || ch == 'y' || ch == 'z') {
```
Comment (rettinghaus) [inline src/iovolpiano.cpp:107]: maybe make this a static set and use `contains` method from C++20?

Comment (lpugin) [reply src/iovolpiano.cpp:107]: Since this fits on one single line I think it is okay to leave it like this.

**Pull request merged**